### ISSUE TITLE
Update Installer before Installing Update

### ISSUE THREAD

**Issue by josegomez** [sqrldev/SQRLDotNetClient#197]
Because the Installer / Updater is included in the zip file with the latest version of the software. When an update is detected, the OLD installer is used to install the latest update (and update the installer)

The issue there is that if there are changes in the new installer that we need applied that won't happen until the next time the installer/updater is run.

We need to find a way to FIRST update the installer/updater before we get and install the latest updates.

Here's my Proposal:
This won't fix the next update for the reasons stated above but it should fix any issues going forward and it doesn't require a whole new program or changes to the Client itself.

The installer today downloads the latest release in a zip file which contains a newer copy of the installer itself.

My proposal is to keep the mechanism as is until that point then change the behavior as follows:

- Installer downloads the latest Zip (as it does today)
- Extract the new Installer from that latest Zip and copy it to %TEMP%
- Then the old installer Process Starts the new installer and passes in:
- the current Install Location (if available) as a param
- the location of the downloaded zip (so we don't download it again) as a param
- and a flag to indicate that it should pick up where it left off
- old installer closes and lets the new installer pick up where it left off (extracting the zip file and doing the install logic)

So basically all it  just uses the old installer to download the newer copy (along with the update as it does today) but then allows the new (better) version to do the actual install.

One advantage of doing it this way is that we don't end up having to download the update multiple times, or separately download the installer which is a self contained executable which would double our download size and time.

Thoughts? or other options are welcome.

**Comment by alexhauser** [comment]
Sounds like a viable plan, @josegomez!

I can't think of any "better" approach right now, so if you want to go ahead with this, please feel free!

We'll just need to make sure that downloading and unzipping is the very first thing the installer does, and that it leaves all other steps to its successor.

Also, we should agree on a common way of handling command line arguments for the installer. At the moment, we're passing in the installation path without any argument specifier, but have one for uninstalling. I guess we should use command line switches for each of the "parameters" we pass in, so that it doesn't get messy.

e.g.

`SQRLPlatformAwareInstaller -uninstall`
`SQRLPlatformAwareInstaller -installPath="C:\Program Files\SQRL"` 
`SQRLPlatformAwareInstaller -archivePath="C:\Users\Test\AppData\Local\Temp\asfdkjk3.zip"`

or some combination of any of them

`SQRLPlatformAwareInstaller -installPath="C:\Test" -archivePath="C:\Temp\asfdkjk3.zip"`

Of course we could also use `--` or `/` as the switch prefix.

**Comment by alexhauser** [comment]
I'll work on this one next, @josegomez.

**Comment by alexhauser** [comment]
@josegomez, I basically have this working, but there are several problems which make me want to rethink the whole approach:

- It's all a HUGE MESS!
When the client launches the installer after it has found an update, it does so by copying the installer executable to `%TEMP%`. Now, when this "old" installer instance extracts the "new" installer binary from the update archive, it needs to find a different location for the "new" installer binary. I solved this by copying/extracting the installer to a unique subdirectory á la `%TEMP\SQRL\DateTime.Now.Ticks.ToString()\` instead, but this is messy and causes other issues:

- The installer log files aren't located in a well-known place anymore, they are scattered around in those temp subdirs and we'll have a hard time getting any logs from our users.

- The `PolicyKit` stuff needs the installer in a fixed location, to that would force us to copy to plain old `%TEMP%` again if we need to invoke `pkexec`.

- The code is getting obscure and hard to follow.
- It makes debugging MUCH harder than it already is.

What's your take on this? Do you have any ideas for eliminating or at least mitigating the problems described above? Or should we think about a different approach?

**Comment by josegomez** [comment]
Ugh yeah that's a huge mess... 
OK how about the following changes.
Remove the ability to select a version, this just complicates things (always force the latest version) allow a command line switch for our testing where we can request a different version or a "pre-release" version.

Force the Installer Log to a specific location in user space. (once of them special folders?)
Then this flow,

If Installer is Run Directly just run through the process as it does registers all the components and pol kit things (pol kit still looks in TEMP for the file)

Then if the update is triggered from the client Download and unpack the latest installer directly from the client (into Temp) then launch the installer (from temp) and follow through. (If on Linux Polikit Launch it)

I think that's a little cleaner. 
Basically the Installer (first time) is just RUN and it does what it does, on Linux it prompts to elevate via sudo (no changes there except removing the option to pick a version)

Then when an update is found make SQRL Client responsible for downloading and unpacking the latest zip into /TEMP/ then invoke the client via policy kit (at all) and run through the install process.

Last but not least, move the install log to a well known and pre-defined location for everything.

Thoughts?

**Comment by alexhauser** [comment]
> Remove the ability to select a version, this just complicates things (always force the latest version) allow a command line switch for our testing where we can request a different version or a "pre-release" version.

This is (somewhat) taken care of already. I've implemented a `-v` or `-version` command line switch which gets passed to the installer and it will pre-select this release. Also, upon fetching releases from Github, the installer now writes the fetched releases into a json file in the temp directory. The "new" installer instance then fetches the releases from this file rather than having to go online again, so that the installation can continue without interruption.

>Force the Installer Log to a specific location in user space. (once of them special folders?)

Yes, this would be the right way to do it anyway I guess.

>Then this flow,

>If Installer is Run Directly just run through the process as it does registers all the components and pol kit things (pol kit still looks in TEMP for the file)

>Then if the update is triggered from the client Download and unpack the latest installer directly from the client (into Temp) then launch the installer (from temp) and follow through. (If on Linux Polikit Launch it)

>I think that's a little cleaner.

Yes, this sounds much cleaner indeed. I like it. It's kinda going back to how we envisioned it in the beginning.

Thanks for your input @josegomez, I'll try to implement it as suggested.
